azkaban-common/src/main/java/azkaban/trigger/builtin/SlaChecker.java

```diff
@@ -32,210 +32,184 @@
-            Utils.parsePeriodString((String) slaOption.getInfo().get(
-                SlaOption.INFO_DURATION));
-        DateTime startTime = new DateTime(node.getStartTime());
-        DateTime nextCheckTime = startTime.plus(duration);
-        this.checkTime = nextCheckTime.getMillis();
+      // Start time needs to be initialized if unset
+      if (startTime < 0) {
+        long t = getStartTime();
+        if (t > 0) {
+          this.startTime = t;
+          ReadablePeriod duration = Utils.parsePeriodString((String) slaOption.getInfo().get(SlaOption.INFO_DURATION));
+          this.slaExpireTime = new DateTime(startTime).plus(duration).getMillis();
+        }
       }
-      status = node.getStatus();
-      return isJobSucceeded(status);
+    } catch (ExecutorManagerException e) {
+      // Log errors but don't crash
+      logger.error("Can't get executable flow.", e);
+      return false;
     }
-    return Boolean.FALSE;
+    return true;
+  }
+
+  private boolean hasSlaExpired() {
+    return slaExpireTime < DateTime.now().getMillis();
   }
 
   // return true to trigger sla action
   @Override
   public Object eval() {
     logger.info("Checking sla for execution " + execId);
-    ExecutableFlow flow;
-    try {
-      flow = executorManager.getExecutableFlow(execId);
-    } catch (ExecutorManagerException e) {
-      logger.error("Can't get executable flow.", e);
-      e.printStackTrace();
-      // something wrong, send out alerts
-      return Boolean.TRUE;
-    }
-    return isSlaMissed(flow);
+    return isSlaFailed();
   }
 
+  /**
+   * TODO Remove convoluted logic. Investigate the need for 2 separate methods for pass and fail test
+   * DO NOT DELETE METHOD. This is used by some weird reflection code using JEXL expression
+   *
+   * @return
+   */
+  @SuppressWarnings("unused")
   public Object isSlaFailed() {
-    ExecutableFlow flow;
-    try {
-      flow = executorManager.getExecutableFlow(execId);
-    } catch (ExecutorManagerException e) {
-      logger.error("Can't get executable flow.", e);
-      // something wrong, send out alerts
-      return Boolean.TRUE;
+    if (!init()) {
+      // Don't trigger SLA actions if unable to init
+      return false;
+    }
+    if (startTime < 0) {
+      return false;
     }
-    return isSlaMissed(flow);
+
+    /*
+     * Fire the trigger (return true) if the desired terminal state isn't reached when
+     *  - the SLA time has already expired, OR
+     *  - the flow / job has already reached a terminal state.
+     */
+    Status status = getStatus();
+    return (isTerminalState(status) || hasSlaExpired()) && !hasReachedDesiredTerminalState();
   }
 
+  /**
+   * TODO Remove convoluted logic. Investigate the need for 2 separate methods for pass and fail test
+   * DO NOT DELETE METHOD. This is used by some weird reflection code using JEXL expression
+   *
+   * @return
+   */
+  @SuppressWarnings("unused")
   public Object isSlaPassed() {
```

Comment:
Please see my comment above.

The APIs are kind of weird. there is a passed as well as failed API and they vary slightly. It is a quite confusing.